File: algorithms/[794]有效的井字游戏/solution.ts
// topics = ["数学"]

function validTicTacToe(board: string[]): boolean {
  const xWin = win(board, 'X');
  const oWin = win(board, 'O');

  let xNums = 0;
  let oNums = 0;
  board
    .join('')
    .split('')
    .forEach(v => {
      if (v == 'X') {
        xNums += 1;
      } else if (v == 'O') {
        oNums += 1;
      }
    });

  return (!oWin && xNums - oNums == 1) || (!xWin && xNums == oNums);
}

function win(board: string[], s: string): boolean {
  for (let i = 0; i < 3; i += 1) {
    if (board[0][i] === s && board[1][i] === s && board[2][i] === s) {
      return true;
    }
    if (board[i][0] === s && board[i][1] === s && board[i][2] === s) {
      return true;
    }
  }

  if (board[0][0] === s && board[1][1] === s && board[2][2] === s) {
    return true;
  }

  if (board[2][0] === s && board[1][1] === s && board[0][2] === s) {
    return true;
  }

  return false;
}

File: algorithms/[794]有效的井字游戏/solution.py
# topics = ["数学"]

from typing import List


class Solution:
    def validTicTacToe(self, board: List[str]) -> bool:
        x_count = sum(row.count('X') for row in board)
        o_count = sum(row.count('O') for row in board)

        def win(player: str) -> bool:
            for i in range(3):
                if all(board[i][j] == player for j in range(3)):
                    return True
                if all(board[j][i] == player for j in range(3)):
                    return True

            if board[1][1] == board[0][0] == board[2][2] == player:
                return True

            if board[1][1] == board[0][2] == board[2][0] == player:
                return True

            return False

        return (not win('X') and x_count == o_count) or (not win('O') and x_count - o_count == 1)
